frame/staking/src/benchmarking.rs

```diff
@@ -137,6 +141,131 @@ pub fn create_validator_with_nominators<T: Config>(
 	Ok((v_stash, nominators))
 }
 
+/// Data for a rebag scenario.
+struct RebagScenario<T: Config> {
+	dest1_stash: T::AccountId,
+	src1_stash: T::AccountId,
+	/// Stash that is expected to be rebagged.
+	src2_stash: T::AccountId,
+	/// Controller of the Stash that is expected to be rebagged.
+	src2_controller: T::AccountId,
+	src3_stash: T::AccountId,
+	src_bag_thresh: BalanceOf<T>,
+	dest_bag_thresh: BalanceOf<T>,
+}
+
+impl<T: Config> RebagScenario<T> {
+	/// Verify that the rebag worked.
+	fn verify_post_rebag(&self) -> Result<(), &'static str> {
+		let RebagScenario {
+			dest1_stash,
+			src1_stash,
+			src2_stash,
+			src3_stash,
+			src_bag_thresh,
+			dest_bag_thresh,
+			..
+		} = self;
+
+		let src2_node = Node::<T>::from_id(&src2_stash).ok_or("node not found for src stash")?;
+		let weight_of = Staking::<T>::weight_of_fn();
+		ensure!(
+			!src2_node.is_misplaced(&weight_of),
+			"src2_node must be in proper place after rebag"
+		);
+
+		let total_issuance = T::Currency::total_issuance();
+
+		let dest_bag_thresh = T::CurrencyToVote::to_vote(*dest_bag_thresh, total_issuance);
+		let dest_bag_ids = Bag::<T>::get(dest_bag_thresh)
+			.ok_or("destination bag not found")?
+			.iter()
+			.map(|n| n.voter().id.clone())
+			.collect::<Vec<_>>();
+		let expected_dest_ids = vec![dest1_stash.clone(), src2_stash.clone()];
+		ensure!(
+			dest_bag_ids == expected_dest_ids,
+			"destination bag is not in expected state after rebag"
+		);
+
+		let src_bag_thresh = T::CurrencyToVote::to_vote(*src_bag_thresh, total_issuance);
+		let src_bag_ids = Bag::<T>::get(src_bag_thresh)
+			.ok_or("source bag not found")?
+			.iter()
+			.map(|n| n.voter().id.clone())
+			.collect::<Vec<_>>();
+		let expected_src_ids = vec![src1_stash.clone(), src3_stash.clone()];
+		ensure!(src_bag_ids == expected_src_ids, "source bag not in expected state after rebag");
+
+		Ok(())
+	}
+}
+
+// Build `RebagScenario`.
+fn build_rebag_scenario<T: Config>(
+	src_bag_thresh: BalanceOf<T>,
+	dest_bag_thresh: BalanceOf<T>,
+) -> Result<RebagScenario<T>, &'static str> {
+	// The most expensive case for this rebag-ing:
```

Comment:
note that complexity of the head/tail being rebagged as well is the same, since although it needs to touch only one neighboar, it will need to update the `Bags` structure as well to update the head/tail of the bag. 

This is also a great test case, can you make sure we cover it? 

```
fn rebag_head_works() {
	// rebagging the head of a bag results in the old bag having a new head and an overall correct state. 
}

```

Reply:
> it will need to update the Bags structure as well to update the head/tail of the bag.

Wether or not the head/tail is updated we still do `bag.put()` after the node is removed, so I would assume we still have to write the bag (assuming the overlay is not smart enough to see it hasn't changed.)

Reply:
```
fn rebag_head_works() {
	// rebagging the head of a bag results in the old bag having a new head and an overall correct state. 
}
```
Added in code TODO for this